clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1416,6 +1415,7 @@ public void assign(Collection<TopicPartition> partitions) {
             if (partitions == null) {
                 throw new IllegalArgumentException("Topic partitions collection to assign to cannot be null");
             }
+            subscriptions.validateAssignFromUserSubscriptionType();
```

Comment:
do we need this validation here? This is already validated when the actual assignment attempt happens (when we call subscriptionState.assignFromUser). I wonder if we could reuse that, and complete the event exceptionally if that call to assignFromUser fails? 

I expect that then the addAndGet call here on ln 1448 would propagate the exception, so we would achieve the result we want, reusing the validation of the subscriptionState, and using the event as a way to propagate the result (successful or failed). Would that work?